src/rules/no-missing-label-refs.js

```diff
@@ -21,106 +21,67 @@ import { findOffsets, illegalShorthandTailPattern } from "../util.js";
 // Helpers
 //-----------------------------------------------------------------------------
 
-const labelPatterns = [
-	// [foo][bar]
-	/\]\[([^\]]+)\]/u,
-
-	// [foo][]
-	/(\]\[\])/u,
-
-	// [foo]
-	/\[([^\]]+)\]/u,
-];
-
-const shorthandTailPattern = /\]\[\]$/u;
-
 /**
  * Finds missing references in a node.
  * @param {TextNode} node The node to check.
- * @param {string} docText The text of the node.
+ * @param {string} nodeText The text of the node.
  * @returns {Array<{label:string,position:Position}>} The missing references.
  */
-function findMissingReferences(node, docText) {
+function findMissingReferences(node, nodeText) {
 	const missing = [];
-	let startIndex = 0;
-	const offset = node.position.start.offset;
 	const nodeStartLine = node.position.start.line;
 	const nodeStartColumn = node.position.start.column;
 
-	/*
-	 * This loop works by searching the string inside the node for the next
-	 * label reference. If there is, it reports an error.
-	 * It then moves the start index to the end of the label reference and
-	 * continues searching the text until the end of the text is found.
-	 */
-	while (startIndex < node.value.length) {
-		const value = node.value.slice(startIndex);
-
-		const match = labelPatterns.reduce((previous, pattern) => {
-			if (previous) {
-				return previous;
-			}
-
-			return value.match(pattern);
-		}, null);
-
-		// check for array instead of null to appease TypeScript
-		if (!Array.isArray(match)) {
-			break;
-		}
+	const labelPattern = /\[(?<left>[^\]]*)\](?:\[(?<right>[^\]]*)\])?/dgu;
```

Comment:
Can we pull this regex out into the module scope? There doesn't seem to be anything context-sensitive that would require it to be in here. I'd rather parse regexes once vs. every time the function is called.

Also, it would be helpful to add a comment explaining what the regex does and how it is used.

Reply:
> Also, it would be helpful to add a comment explaining what the regex does and how it is used.

Added in https://github.com/eslint/markdown/pull/290/commits/60d5735877e4aeea7569c5439541df0d285d24df.

Reply:
> Can we pull this regex out into the module scope? There doesn't seem to be anything context-sensitive that would require it to be in here. I'd rather parse regexes once vs. every time the function is called.

The `g` flag makes it mutable - each `.exec()` modifies `lastIndex`. I think it's a good practice to keep those regex instances scoped to the search instance that's using them, especially if they are used in a loop, because a small change in the code, like adding an early exit in the loop, can cause bugs that are difficult to reproduce and figure out.